src/app/Events/EventTypes.tsx

```diff
@@ -0,0 +1,140 @@
+}
+
+export interface OptionDescriptor {
+  name: string;
+  description: string;
+  defaultValue: string;
+}
+
+export const EventTypes = (props) => {
+  const context = React.useContext(ServiceContext);
+
+  const [types, setTypes] = React.useState([] as EventType[]);
+  const [displayedTypes, setDisplayedTypes] = React.useState([] as any[]);
+  const [currentPage, setCurrentPage] = React.useState(1);
+  const [perPage, setPerPage] = React.useState(10);
+  const [openRow, setOpenRow] = React.useState(-1);
+  const [filterText, setFilterText] = React.useState('');
+
+  const tableColumns = [
+    {
+      title: 'Name',
+      cellFormatters: [expandable]
+    },
+    'Type ID',
+    'Description',
+    'Categories'
+  ];
+
+  React.useEffect(() => {
+    const sub = context.commandChannel.onResponse('list-event-types')
+      .pipe(
+        filter(m => m.status === 0),
+        map(m => m.payload),
+      )
+      .subscribe(types => setTypes(types));
+    return () => sub.unsubscribe();
+  }, []);
+
+  React.useEffect(() => {
+    context.commandChannel.sendMessage('list-event-types');
+  }, []);
+
+  React.useEffect(() => {
+    const offset = (currentPage - 1) * perPage;
+    const page = filterTypesByText(types, filterText).slice(offset, offset + perPage);
+
+    const rows: any[] = [];
+    page.forEach((t: EventType, idx: number) => {
+      rows.push({ cells: [ t.name, t.typeId, t.description, t.category.join(', ').trim() ], isOpen: (idx === openRow) });
+      if (idx === openRow) {
+        let child = '';
+        for (const opt in t.options) {
+          child += `${opt}=[${t.options[opt].defaultValue}]\t`;
+        }
+        rows.push({ parent: idx, fullWidth: true, cells: [ child ] });
+      }
+    });
+
+    setDisplayedTypes(rows);
+  }, [currentPage, perPage, types, openRow, filterText]);
+
+  const filterTypesByText = (types, filter) => {
+    if (!filter) {
+      return types;
+    }
+    const includesSubstr = (a, b) => !!a && !!b && a.toLowerCase().includes(b.trim().toLowerCase());
+    return types.filter(t => {
+      if (includesSubstr(t.name, filter)) {
+        return true;
+      }
+      if (includesSubstr(t.typeId, filter)) {
+        return true;
+      }
+      if (includesSubstr(t.description, filter)) {
+        return true;
+      }
+      if (t.category.some(c => includesSubstr(c, filter))) {
+        return true;
+      }
+      return false
+    });
+  };
+
+  const onCurrentPage = (evt, currentPage) => {
+    setOpenRow(-1);
+    setCurrentPage(currentPage);
+  };
+
+  const onPerPage = (evt, perPage) => setPerPage(perPage);
+
+  const onCollapse = (event, rowKey, isOpen) => {
+    if (isOpen) {
+      if (openRow === -1) {
+        setOpenRow(rowKey);
+      } else {
+        setOpenRow(rowKey > openRow ? rowKey - 1 : rowKey);
+      }
+    } else {
+      setOpenRow(-1);
+    }
+  };
+
+  // TODO replace table with data list so collapsed event options can be custom formatted
```

Comment:
Is this TODO still relevant? I'd like to see an issue updated about this, or filed for this if appropriate. Maybe this TODO can link the issue, up to you for that part.

Reply:
It is still relevant. I'll file a new issue for it, along with start working on a mockup of what the expanded rows could look like.